File: backend/routes/quests.js
const express = require('express');
const router = express.Router();
const { pool } = require('../config/database');

router.get('/game/:gameId', async (req, res, next) => {
  try {
    const { type, difficulty } = req.query;
    
    let query = 'SELECT * FROM quests WHERE game_id = $1';
    const params = [req.params.gameId];
    let paramCount = 2;

    if (type) {
      query += ` AND quest_type = $${paramCount}`;
      params.push(type);
      paramCount++;
    }

    if (difficulty) {
      query += ` AND difficulty_rating <= $${paramCount}`;
      params.push(parseInt(difficulty));
      paramCount++;
    }

    query += ' ORDER BY quest_type, name';

    const result = await pool.query(query, params);
    res.json({ quests: result.rows });
  } catch (error) {
    next(error);
  }
});

router.get('/:id', async (req, res, next) => {
  try {
    const result = await pool.query(
      'SELECT * FROM quests WHERE id = $1',
      [req.params.id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Quest not found' });
    }

    res.json({ quest: result.rows[0] });
  } catch (error) {
    next(error);
  }
});

router.get('/user/progress', async (req, res, next) => {
  try {
    const { game_id, status } = req.query;
    
    let query = `
      SELECT uqp.*, q.name, q.description, q.quest_type, q.difficulty_rating, g.name as game_name
      FROM user_quest_progress uqp
      JOIN quests q ON uqp.quest_id = q.id
      JOIN games g ON q.game_id = g.id
      WHERE uqp.user_id = $1
    `;
    const params = [req.user.userId];
    let paramCount = 2;

    if (game_id) {
      query += ` AND q.game_id = $${paramCount}`;
      params.push(game_id);
      paramCount++;
    }

    if (status) {
      query += ` AND uqp.status = $${paramCount}`;
      params.push(status);
      paramCount++;
    }

    query += ' ORDER BY uqp.started_at DESC';

    const result = await pool.query(query, params);
    res.json({ progress: result.rows });
  } catch (error) {
    next(error);
  }
});

router.post('/:id/start', async (req, res, next) => {
  try {
    const questId = req.params.id;
    const userId = req.user.userId;

    const questCheck = await pool.query(
      'SELECT id, name FROM quests WHERE id = $1',
      [questId]
    );

    if (questCheck.rows.length === 0) {
      return res.status(404).json({ error: 'Quest not found' });
    }

    const result = await pool.query(
      `INSERT INTO user_quest_progress (user_id, quest_id, status, started_at)
       VALUES ($1, $2, 'in_progress', NOW())
       ON CONFLICT (user_id, quest_id) 
       DO UPDATE SET status = 'in_progress', started_at = NOW()
       RETURNING *`,
      [userId, questId]
    );

    res.json({
      message: 'Quest started',
      progress: result.rows[0],
      quest: questCheck.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

router.put('/:id/progress', async (req, res, next) => {
  try {
    const { current_objective, status } = req.body;
    const questId = req.params.id;
    const userId = req.user.userId;

    const result = await pool.query(
      `UPDATE user_quest_progress 
       SET current_objective = COALESCE($1, current_objective),
           status = COALESCE($2, status),
           completed_at = CASE WHEN $2 = 'completed' THEN NOW() ELSE completed_at END
       WHERE user_id = $3 AND quest_id = $4
       RETURNING *`,
      [current_objective, status, userId, questId]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Quest progress not found' });
    }

    res.json({
      message: 'Progress updated',
      progress: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

router.post('/:id/complete', async (req, res, next) => {
  try {
    const questId = req.params.id;
    const userId = req.user.userId;

    const progressResult = await pool.query(
      'SELECT started_at FROM user_quest_progress WHERE user_id = $1 AND quest_id = $2',
      [userId, questId]
    );

    if (progressResult.rows.length === 0) {
      return res.status(404).json({ error: 'Quest not started' });
    }

    const startTime = progressResult.rows[0].started_at;
    const completionTime = Math.floor((Date.now() - new Date(startTime).getTime()) / 60000); // minutes

    const result = await pool.query(
      `UPDATE user_quest_progress 
       SET status = 'completed',
           completed_at = NOW(),
           completion_time = $1
       WHERE user_id = $2 AND quest_id = $3
       RETURNING *`,
      [completionTime, userId, questId]
    );

    res.json({
      message: 'Quest completed! 🎉',
      progress: result.rows[0],
      completion_time_minutes: completionTime
    });
  } catch (error) {
    next(error);
  }
});

router.get('/:id/stats', async (req, res, next) => {
  try {
    const questId = req.params.id;

    const stats = await pool.query(
      `SELECT 
         COUNT(*) as total_attempts,
         COUNT(CASE WHEN status = 'completed' THEN 1 END) as completions,
         AVG(completion_time) FILTER (WHERE status = 'completed') as avg_completion_time,
         AVG(hints_used) as avg_hints_used
       FROM user_quest_progress
       WHERE quest_id = $1`,
      [questId]
    );

    res.json({ stats: stats.rows[0] });
  } catch (error) {
    next(error);
  }
});

router.post('/:id/abandon', async (req, res, next) => {
  try {
    const result = await pool.query(
      `UPDATE user_quest_progress 
       SET status = 'abandoned'
       WHERE user_id = $1 AND quest_id = $2
       RETURNING *`,
      [req.user.userId, req.params.id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Quest progress not found' });
    }

    res.json({
      message: 'Quest abandoned',
      progress: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

module.exports = router;

File: backend/routes/quests.py
# routes/quests.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from utils.database import db
from models.quest import Quest
from models.game import Game

quests_bp = Blueprint('quests', __name__)

@quests_bp.route('', methods=['POST'])
@jwt_required()
def create_quest():
    """Create a new quest"""
    try:
        data = request.json
        user_id = get_jwt_identity()
        
        # Validate required fields
        if not data.get('game_id'):
            return jsonify({'error': 'game_id is required'}), 400
        
        if not data.get('name') and not data.get('title'):
            return jsonify({'error': 'Quest name is required'}), 400
        
        # Check if game exists
        game = Game.query.get(data.get('game_id'))
        if not game:
            return jsonify({'error': 'Game not found'}), 404
        
        # Create new quest
        new_quest = Quest(
            game_id=data.get('game_id'),
            name=data.get('name') or data.get('title'),
            description=data.get('description', ''),
            difficulty=data.get('difficulty', 5),
            status='active',
            completed=False
        )
        
        db.session.add(new_quest)
        db.session.commit()
        
        return jsonify({
            'message': 'Quest created successfully',
            'quest': {
                'id': new_quest.id,
                'name': new_quest.name,
                'description': new_quest.description,
                'difficulty': new_quest.difficulty,
                'status': new_quest.status,
                'completed': new_quest.completed,
                'game_id': new_quest.game_id
            }
        }), 201
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@quests_bp.route('/game/<int:game_id>', methods=['GET'])
@jwt_required()
def get_game_quests(game_id):
    """Get all quests for a specific game"""
    try:
        user_id = get_jwt_identity()
        
        # Check if game exists
        game = Game.query.get(game_id)
        if not game:
            return jsonify({'error': 'Game not found'}), 404
        
        # Get all quests for this game
        quests = Quest.query.filter_by(game_id=game_id).all()
        
        return jsonify({
            'quests': [{
                'id': q.id,
                'name': q.name,
                'description': q.description,
                'difficulty': q.difficulty,
                'status': q.status,
                'completed': q.completed,
                'game_id': q.game_id,
                'created_at': q.created_at.isoformat() if hasattr(q, 'created_at') and q.created_at else None
            } for q in quests]
        }), 200
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@quests_bp.route('/<int:quest_id>', methods=['GET'])
@jwt_required()
def get_quest(quest_id):
    """Get a single quest"""
    try:
        user_id = get_jwt_identity()
        quest = Quest.query.get_or_404(quest_id)
        
        return jsonify({
            'quest': {
                'id': quest.id,
                'name': quest.name,
                'description': quest.description,
                'difficulty': quest.difficulty,
                'status': quest.status,
                'completed': quest.completed,
                'game_id': quest.game_id
            }
        }), 200
        
    except Exception as e:
        return jsonify({'error': str(e)}), 404

@quests_bp.route('/<int:quest_id>', methods=['PATCH'])
@jwt_required()
def update_quest(quest_id):
    """Update a quest"""
    try:
        user_id = get_jwt_identity()
        quest = Quest.query.get_or_404(quest_id)
        data = request.json
        
        # Update fields if provided
        if 'name' in data:
            quest.name = data['name']
        if 'description' in data:
            quest.description = data['description']
        if 'difficulty' in data:
            quest.difficulty = data['difficulty']
        if 'status' in data:
            quest.status = data['status']
        if 'completed' in data:
            quest.completed = data['completed']
        
        db.session.commit()
        
        return jsonify({
            'message': 'Quest updated successfully',
            'quest': {
                'id': quest.id,
                'name': quest.name,
                'description': quest.description,
                'difficulty': quest.difficulty,
                'status': quest.status,
                'completed': quest.completed
            }
        }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@quests_bp.route('/<int:quest_id>', methods=['DELETE'])
@jwt_required()
def delete_quest(quest_id):
    """Delete a quest"""
    try:
        user_id = get_jwt_identity()
        quest = Quest.query.get_or_404(quest_id)
        
        db.session.delete(quest)
        db.session.commit()
        
        return jsonify({'message': 'Quest deleted successfully'}), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@quests_bp.route('/<int:quest_id>/complete', methods=['POST'])
@jwt_required()
def complete_quest(quest_id):
    """Mark a quest as complete"""
    try:
        user_id = get_jwt_identity()
        quest = Quest.query.get_or_404(quest_id)
        
        quest.completed = True
        quest.status = 'completed'
        
        db.session.commit()
        
        return jsonify({
            'message': 'Quest completed! 🎉',
            'quest': {
                'id': quest.id,
                'name': quest.name,
                'completed': quest.completed,
                'status': quest.status
            }
        }), 200
        
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@quests_bp.route('/all', methods=['GET'])
@jwt_required()
def get_all_quests():
    """Get all quests across all games for the current user"""
    try:
        user_id = get_jwt_identity()
        
        # Get all games for this user
        games = Game.query.filter_by(user_id=user_id).all()
        game_ids = [g.id for g in games]
        
        # Get all quests for these games
        quests = Quest.query.filter(Quest.game_id.in_(game_ids)).all()
        
        return jsonify({
            'quests': [{
                'id': q.id,
                'name': q.name,
                'description': q.description,
                'difficulty': q.difficulty,
                'status': q.status,
                'completed': q.completed,
                'game_id': q.game_id
            } for q in quests]
        }), 200
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500